File: src/main/java/org/moinex/ui/dialog/investment/BondTransactionsController.java
/*
 * Filename: BondTransactionsController.java
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import java.math.BigDecimal;
import java.util.List;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.fxml.FXML;
import javafx.geometry.Pos;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.OperationType;
import org.moinex.model.investment.BondOperation;
import org.moinex.service.BondService;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class BondTransactionsController {
    @FXML private TableView<BondOperation> operationTableView;

    @FXML private TextField searchField;

    private ConfigurableApplicationContext springContext;

    private List<BondOperation> operations;

    private BondService bondService;
    private I18nService i18nService;

    private final String NA_ITEM = "N/A";

    @Autowired
    public BondTransactionsController(
            BondService bondService,
            ConfigurableApplicationContext springContext,
            I18nService i18nService) {
        this.bondService = bondService;
        this.springContext = springContext;
        this.i18nService = i18nService;
    }

    @FXML
    public void initialize() {
        loadOperationsFromDatabase();

        configureOperationTableView();

        updateOperationTableView();

        searchField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            updateOperationTableView();
                        });
    }

    @FXML
    private void handleCancel() {
        Stage stage = (Stage) searchField.getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleEditOperation() {
        BondOperation selectedOperation = operationTableView.getSelectionModel().getSelectedItem();

        if (selectedOperation == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_DIALOG_NO_OPERATION_SELECTED_EDIT_MESSAGE));
            return;
        }

        if (selectedOperation.getOperationType() == OperationType.BUY) {
            WindowUtils.openModalWindow(
                    Constants.EDIT_BOND_PURCHASE_FXML,
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EDIT_PURCHASE_TITLE),
                    springContext,
                    (EditBondPurchaseController controller) ->
                            controller.setOperation(selectedOperation),
                    List.of(
                            () -> {
                                loadOperationsFromDatabase();
                                updateOperationTableView();
                            }));
        } else {
            WindowUtils.openModalWindow(
                    Constants.EDIT_BOND_SALE_FXML,
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EDIT_SALE_TITLE),
                    springContext,
                    (EditBondSaleController controller) ->
                            controller.setOperation(selectedOperation),
                    List.of(
                            () -> {
                                loadOperationsFromDatabase();
                                updateOperationTableView();
                            }));
        }
    }

    @FXML
    private void handleDeleteOperation() {
        BondOperation selectedOperation = operationTableView.getSelectionModel().getSelectedItem();

        if (selectedOperation == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .BOND_DIALOG_NO_OPERATION_SELECTED_DELETE_MESSAGE));
            return;
        }

        String operationType =
                UIUtils.translateOperationType(selectedOperation.getOperationType(), i18nService);
        String bondName = selectedOperation.getBond().getName();
        String symbol = selectedOperation.getBond().getSymbol();
        String bondDisplay =
                bondName + (symbol != null && !symbol.isBlank() ? " (" + symbol + ")" : "");

        String message =
                i18nService
                        .tr(Constants.TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_OPERATION_MESSAGE)
                        .replace("{operationType}", operationType)
                        .replace("{bond}", bondDisplay);

        if (WindowUtils.showConfirmationDialog(
                i18nService.tr(
                        Constants.TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_OPERATION_TITLE),
                message,
                i18nService.getBundle())) {
            try {
                bondService.deleteOperation(selectedOperation.getId());
                loadOperationsFromDatabase();
                updateOperationTableView();

                WindowUtils.showSuccessDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_OPERATION_DELETED_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_OPERATION_DELETED_MESSAGE));
            } catch (Exception e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(Constants.TranslationKeys.DIALOG_ERROR_TITLE),
                        e.getMessage());
            }
        }
    }

    private void loadOperationsFromDatabase() {
        operations = bondService.getAllOperations();
    }

    private void updateOperationTableView() {
        String similarTextOrId = searchField.getText().toLowerCase();

        operationTableView.getItems().clear();

        if (similarTextOrId.isEmpty()) {
            operationTableView.getItems().setAll(operations);
        } else {
            operations.stream()
                    .filter(
                            op -> {
                                String id = op.getId().toString();
                                String bondName = op.getBond().getName().toLowerCase();
                                String bondSymbol =
                                        op.getBond().getSymbol() != null
                                                ? op.getBond().getSymbol().toLowerCase()
                                                : "";
                                String date =
                                        UIUtils.formatDateForDisplay(
                                                op.getWalletTransaction().getDate(), i18nService);
                                String quantity = op.getQuantity().toString();
                                String unitPrice = op.getUnitPrice().toString();
                                String fees = op.getFees() != null ? op.getFees().toString() : "";
                                String taxes =
                                        op.getTaxes() != null ? op.getTaxes().toString() : "";
                                String operationType = op.getOperationType().name().toLowerCase();
                                String walletName =
                                        op.getWalletTransaction() != null
                                                ? op.getWalletTransaction()
                                                        .getWallet()
                                                        .getName()
                                                        .toLowerCase()
                                                : "";

                                return id.contains(similarTextOrId)
                                        || bondName.contains(similarTextOrId)
                                        || bondSymbol.contains(similarTextOrId)
                                        || date.contains(similarTextOrId)
                                        || quantity.contains(similarTextOrId)
                                        || unitPrice.contains(similarTextOrId)
                                        || fees.contains(similarTextOrId)
                                        || taxes.contains(similarTextOrId)
                                        || operationType.contains(similarTextOrId)
                                        || walletName.contains(similarTextOrId);
                            })
                    .forEach(operationTableView.getItems()::add);
        }

        operationTableView.refresh();
    }

    private void configureOperationTableView() {
        TableColumn<BondOperation, Integer> idColumn = getOperationIdTableColumn();

        TableColumn<BondOperation, String> operationTypeColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.BOND_TABLE_OPERATION_TYPE));
        operationTypeColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.translateOperationType(
                                        param.getValue().getOperationType(), i18nService)));

        TableColumn<BondOperation, String> bondNameColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_BOND));
        bondNameColumn.setCellValueFactory(
                param -> {
                    String symbol = param.getValue().getBond().getSymbol();
                    return new SimpleStringProperty(
                            param.getValue().getBond().getName()
                                    + (symbol != null && !symbol.isBlank()
                                            ? " (" + symbol + ")"
                                            : ""));
                });

        TableColumn<BondOperation, String> bondTypeColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_TYPE));
        bondTypeColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.translateBondType(
                                        param.getValue().getBond().getType(), i18nService)));

        TableColumn<BondOperation, String> dateColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_DATE));
        dateColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.formatDateForDisplay(
                                        param.getValue().getWalletTransaction().getDate(),
                                        i18nService)));

        TableColumn<BondOperation, String> quantityColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_QUANTITY));
        quantityColumn.setCellValueFactory(
                param -> new SimpleObjectProperty<>(param.getValue().getQuantity().toString()));

        TableColumn<BondOperation, String> unitPriceColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_UNIT_PRICE));
        unitPriceColumn.setCellValueFactory(
                param ->
                        new SimpleObjectProperty<>(
                                UIUtils.formatCurrency(param.getValue().getUnitPrice())));

        TableColumn<BondOperation, String> feesColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_FEES));
        feesColumn.setCellValueFactory(
                param ->
                        new SimpleObjectProperty<>(
                                param.getValue().getFees() != null
                                        ? UIUtils.formatCurrency(param.getValue().getFees())
                                        : NA_ITEM));

        TableColumn<BondOperation, String> taxesColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_TAXES));
        taxesColumn.setCellValueFactory(
                param ->
                        new SimpleObjectProperty<>(
                                param.getValue().getTaxes() != null
                                        ? UIUtils.formatCurrency(param.getValue().getTaxes())
                                        : NA_ITEM));

        TableColumn<BondOperation, String> profitLossColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_PROFIT_LOSS));
        profitLossColumn.setCellValueFactory(
                param -> {
                    BondOperation op = param.getValue();
                    if (op.getOperationType() == OperationType.BUY) {
                        return new SimpleObjectProperty<>(NA_ITEM);
                    }
                    BigDecimal profitLoss = bondService.calculateOperationProfitLoss(op);
                    return new SimpleObjectProperty<>(UIUtils.formatCurrencySigned(profitLoss));
                });
        profitLossColumn.setCellFactory(
                column ->
                        new TableCell<>() {
                            @Override
                            protected void updateItem(String item, boolean empty) {
                                super.updateItem(item, empty);
                                if (item == null || empty) {
                                    setText(null);
                                    setStyle("");
                                } else {
                                    setText(item);
                                    getStyleClass()
                                            .removeAll(
                                                    Constants.POSITIVE_BALANCE_STYLE,
                                                    Constants.NEGATIVE_BALANCE_STYLE,
                                                    Constants.NEUTRAL_BALANCE_STYLE);

                                    if (!item.equals(NA_ITEM)) {
                                        if (item.startsWith("+")) {
                                            getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);
                                        } else if (item.startsWith("-")) {
                                            getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);
                                        } else {
                                            getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);
                                        }
                                    }
                                }
                            }
                        });

        TableColumn<BondOperation, String> walletNameColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_WALLET));
        walletNameColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                param.getValue().getWalletTransaction() != null
                                        ? param.getValue()
                                                .getWalletTransaction()
                                                .getWallet()
                                                .getName()
                                        : ""));

        operationTableView.getColumns().add(idColumn);
        operationTableView.getColumns().add(operationTypeColumn);
        operationTableView.getColumns().add(bondNameColumn);
        operationTableView.getColumns().add(bondTypeColumn);
        operationTableView.getColumns().add(dateColumn);
        operationTableView.getColumns().add(quantityColumn);
        operationTableView.getColumns().add(unitPriceColumn);
        operationTableView.getColumns().add(feesColumn);
        operationTableView.getColumns().add(taxesColumn);
        operationTableView.getColumns().add(profitLossColumn);
        operationTableView.getColumns().add(walletNameColumn);
    }

    private TableColumn<BondOperation, Integer> getOperationIdTableColumn() {
        TableColumn<BondOperation, Integer> idColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_ID));
        idColumn.setCellValueFactory(param -> new SimpleObjectProperty<>(param.getValue().getId()));

        idColumn.setCellFactory(
                column ->
                        new TableCell<>() {
                            @Override
                            protected void updateItem(Integer item, boolean empty) {
                                super.updateItem(item, empty);
                                if (item == null || empty) {
                                    setText(null);
                                } else {
                                    setText(item.toString());
                                    setAlignment(Pos.CENTER);
                                    setStyle("-fx-padding: 0;");
                                }
                            }
                        });
        return idColumn;
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/view/BondTransactionsController.kt
/*
 * Filename: BondTransactionsController.kt (original filename: BondTransactionsController.java)
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.view

import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.TableCell
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Files
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.enums.OperationType
import org.moinex.model.investment.BondOperation
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondService
import org.moinex.ui.dialog.investment.update.EditBondPurchaseController
import org.moinex.ui.dialog.investment.update.EditBondSaleController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller

@Controller
class BondTransactionsController(
    private val bondService: BondService,
    private val springContext: ConfigurableApplicationContext,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var operationTableView: TableView<BondOperation>

    @FXML
    private lateinit var searchField: TextField

    private var operations: MutableList<BondOperation> = mutableListOf()

    @FXML
    fun initialize() {
        loadOperationsFromDatabase()
        configureOperationTableView()
        updateOperationTableView()

        searchField.textProperty().addListener { _, _, _ ->
            updateOperationTableView()
        }
    }

    @FXML
    private fun handleCancel() {
        (searchField.scene.window as Stage).close()
    }

    @FXML
    private fun handleEditOperation() {
        val selectedOperation = operationTableView.selectionModel.selectedItem

        if (selectedOperation == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_EDIT_MESSAGE),
            )
            return
        }

        if (selectedOperation.operationType == OperationType.BUY) {
            WindowUtils.openModalWindow(
                Files.EDIT_BOND_PURCHASE_FXML,
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EDIT_PURCHASE_TITLE),
                springContext,
                { controller: EditBondPurchaseController ->
                    controller.setOperation(selectedOperation)
                },
                listOf(
                    Runnable {
                        loadOperationsFromDatabase()
                        updateOperationTableView()
                    },
                ),
            )
        } else {
            WindowUtils.openModalWindow(
                Files.EDIT_BOND_SALE_FXML,
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EDIT_SALE_TITLE),
                springContext,
                { controller: EditBondSaleController ->
                    controller.setOperation(selectedOperation)
                },
                listOf(
                    Runnable {
                        loadOperationsFromDatabase()
                        updateOperationTableView()
                    },
                ),
            )
        }
    }

    @FXML
    private fun handleDeleteOperation() {
        val selectedOperation = operationTableView.selectionModel.selectedItem

        if (selectedOperation == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_OPERATION_SELECTED_DELETE_MESSAGE),
            )
            return
        }

        val operationType = UIUtils.translateOperationType(selectedOperation.operationType)
        val bondName = selectedOperation.bond.name
        val symbol = selectedOperation.bond.symbol
        val bondDisplay = bondName + if (!symbol.isNullOrBlank()) " ($symbol)" else ""

        val message =
            preferencesService
                .translate(TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_OPERATION_MESSAGE)
                .replace("{operationType}", operationType)
                .replace("{bond}", bondDisplay)

        if (WindowUtils.showConfirmationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_OPERATION_TITLE),
                message,
                preferencesService.bundle,
            )
        ) {
            runCatching {
                bondService.deleteBondOperation(selectedOperation.id!!)
                loadOperationsFromDatabase()
                updateOperationTableView()

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_OPERATION_DELETED_TITLE),
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_OPERATION_DELETED_MESSAGE),
                )
            }.onFailure { e ->
                WindowUtils.showErrorDialog(
                    preferencesService.translate(TranslationKeys.DIALOG_ERROR_TITLE),
                    e.message ?: "",
                )
            }
        }
    }

    private fun loadOperationsFromDatabase() {
        operations = bondService.getAllOperations().toMutableList()
    }

    private fun updateOperationTableView() {
        val similarTextOrId = searchField.text.lowercase()

        operationTableView.items.clear()

        if (similarTextOrId.isEmpty()) {
            operationTableView.items.setAll(operations)
        } else {
            operations
                .filter { op ->
                    val id = op.id.toString()
                    val bondName = op.bond.name.lowercase()
                    val bondSymbol = op.bond.symbol?.lowercase() ?: ""
                    val date = UIUtils.formatDateForDisplay(op.walletTransaction!!.date)
                    val quantity = op.quantity.toString()
                    val unitPrice = op.unitPrice.toString()
                    val fees = op.fees.toString()
                    val taxes = op.taxes.toString()
                    val operationType = op.operationType.name.lowercase()
                    val walletName =
                        op.walletTransaction
                            ?.wallet
                            ?.name
                            ?.lowercase() ?: ""

                    id.contains(similarTextOrId) ||
                        bondName.contains(similarTextOrId) ||
                        bondSymbol.contains(similarTextOrId) ||
                        date.contains(similarTextOrId) ||
                        quantity.contains(similarTextOrId) ||
                        unitPrice.contains(similarTextOrId) ||
                        fees.contains(similarTextOrId) ||
                        taxes.contains(similarTextOrId) ||
                        operationType.contains(similarTextOrId) ||
                        walletName.contains(similarTextOrId)
                }.forEach { operationTableView.items.add(it) }
        }

        operationTableView.refresh()
    }

    private fun configureOperationTableView() {
        val idColumn =
            TableColumn<BondOperation, Int>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_ID),
            ).apply {
                setCellValueFactory { param -> SimpleObjectProperty(param.value.id) }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val operationTypeColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_OPERATION_TYPE),
            ).apply {
                setCellValueFactory { param ->
                    SimpleStringProperty(UIUtils.translateOperationType(param.value.operationType))
                }
            }

        val bondNameColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_BOND),
            ).apply {
                setCellValueFactory { param ->
                    val symbol = param.value.bond.symbol
                    SimpleStringProperty(
                        param.value.bond.name + if (!symbol.isNullOrBlank()) " ($symbol)" else "",
                    )
                }
            }

        val bondTypeColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_TYPE),
            ).apply {
                setCellValueFactory { param ->
                    SimpleStringProperty(UIUtils.translateBondType(param.value.bond.type))
                }
            }

        val dateColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_DATE),
            ).apply {
                setCellValueFactory { param ->
                    SimpleStringProperty(UIUtils.formatDateForDisplay(param.value.walletTransaction!!.date))
                }
            }

        val quantityColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_QUANTITY),
            ).apply {
                setCellValueFactory { param -> SimpleObjectProperty(param.value.quantity.toString()) }
            }

        val unitPriceColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_UNIT_PRICE),
            ).apply {
                setCellValueFactory { param ->
                    SimpleObjectProperty(UIUtils.formatCurrency(param.value.unitPrice))
                }
            }

        val feesColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_FEES),
            ).apply {
                setCellValueFactory { param ->
                    SimpleObjectProperty(UIUtils.formatCurrency(param.value.fees))
                }
            }

        val taxesColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_TAXES),
            ).apply {
                setCellValueFactory { param ->
                    SimpleObjectProperty(UIUtils.formatCurrency(param.value.taxes))
                }
            }

        val profitLossColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_PROFIT_LOSS),
            ).apply {
                setCellValueFactory { param ->
                    val op = param.value
                    if (op.operationType == OperationType.BUY) {
                        SimpleObjectProperty(Constants.NA_DATA)
                    } else {
                        SimpleObjectProperty(UIUtils.formatCurrencySigned(op.netProfit))
                    }
                }
                setCellFactory {
                    object : TableCell<BondOperation, String>() {
                        override fun updateItem(
                            item: String?,
                            empty: Boolean,
                        ) {
                            super.updateItem(item, empty)
                            if (item == null || empty) {
                                text = null
                                style = ""
                            } else {
                                text = item
                                styleClass.removeAll(
                                    Styles.POSITIVE_BALANCE_STYLE,
                                    Styles.NEGATIVE_BALANCE_STYLE,
                                    Styles.NEUTRAL_BALANCE_STYLE,
                                )

                                if (item != Constants.NA_DATA) {
                                    when {
                                        item.startsWith("+") -> styleClass.add(Styles.POSITIVE_BALANCE_STYLE)
                                        item.startsWith("-") -> styleClass.add(Styles.NEGATIVE_BALANCE_STYLE)
                                        else -> styleClass.add(Styles.NEUTRAL_BALANCE_STYLE)
                                    }
                                }
                            }
                        }
                    }
                }
            }

        val walletNameColumn =
            TableColumn<BondOperation, String>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_WALLET),
            ).apply {
                setCellValueFactory { param ->
                    SimpleStringProperty(
                        if (param.value.walletTransaction != null) {
                            param.value.walletTransaction!!
                                .wallet.name
                        } else {
                            ""
                        },
                    )
                }
            }

        operationTableView.columns.addAll(
            idColumn,
            operationTypeColumn,
            bondNameColumn,
            bondTypeColumn,
            dateColumn,
            quantityColumn,
            unitPriceColumn,
            feesColumn,
            taxesColumn,
            profitLossColumn,
            walletNameColumn,
        )
    }
}
